### DISCUSSION TITLE
How to access self.log on_init_() or before_train()?

### DISCUSSION THREAD

**Discussion by allanchan339** [Lightning-AI/pytorch-lightning#11641] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
By pl docs, callbacks to implement unnecessary research code are recommended. Therefore, I would like to record the total trainable parameters numbers, total size, and another record in the logging system. 
However, it seems to me that the self.log can only be accessed after training is initialized. 
Any better way to access the log once only before training?

**Comment by ananthsub** [comment]
have you taken a look at the model summary callback? https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.callbacks.ModelSummary.html#pytorch_lightning.callbacks.ModelSummary

that's an example of something which runs before training and can include the model size, parameters per module, and whose output can be logged to another system. by default it prints out a table

You can do something once before training in the `on_train_start` hook.

**Reply by allanchan339** [reply]
Yes 

> have you taken a look at the model summary callback? https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.callbacks.ModelSummary.html#pytorch_lightning.callbacks.ModelSummary
> 
> that's an example of something which runs before training and can include the model size, parameters per module, and whose output can be logged to another system. by default it prints out a table
> 
> You can do something once before training in the `on_train_start` hook.

I can access self.log on on_train_start() hook. However, the logger keeps logging the number of the parameters. Instead of logging as a value, the logger displays a curve to me. If I don't get it wrong, it seems the trainer repeatedly calls the on_train_start() hook instead of single call.

**Reply by allanchan339** [reply]
![W B Chart 27_1_2022 下午3_54_57](https://user-images.githubusercontent.com/38370649/151315274-1b1c2dae-85d3-4b2b-b211-c5d7c3013efe.png)


> have you taken a look at the model summary callback? https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.callbacks.ModelSummary.html#pytorch_lightning.callbacks.ModelSummary
> 
> that's an example of something which runs before training and can include the model size, parameters per module, and whose output can be logged to another system. by default it prints out a table
> 
> You can do something once before training in the `on_train_start` hook.

This is the curve I am talking about.

**Reply by ananthsub** [reply]
Can you use the trainer.logger to log the metrics directly instead of the callback.log function?

**Reply by allanchan339** [reply]
> Can you use the trainer.logger to log the metrics directly instead of the callback.log function?

Yes. However, it seems to violate the recommended method. Also, I would like to prevent calling a global variable `logger` if possible
